### PR TITLE
KAFKA-7396 Materialized, Serialized, Joined, Consumed and Produced with implicit Serdes

### PR DESCRIPTION
**Pull request by joan38**
We want to make sure that we always have a serde for all Materialized, Serialized, Joined, Consumed and Produced.
For that we can make use of the implicit parameters in Scala.

KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-365%3A+Materialized%2C+Serialized%2C+Joined%2C+Consumed+and+Produced+with+implicit+Serde

### CODE REVIEW

**Comment by joan38** [comment]
This one is ready for review @guozhangwang @vvcephei @ijuma @debasishg

**Review by guozhangwang — COMMENTED** [review]
LGTM. @vvcephei could you take another look?

**Comment by debasishg** [comment]
LGTM .. should we also do the same for `Serialized` ?

**Comment by joan38** [comment]
@debasishg good point even tho I don't think we have to create manually any `Serialized` are we?
I guess if we do the `Serialized`, we should also do the `Consumed` and `Produced`?

**Comment by joan38** [comment]
@mjsax Could you label this PR `stream` please?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/package.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/package.scala]: Can you explain why this is necessary? Everything seems to compile and run fine when I remove it.

**Comment by mjsax** [comment]
Wondering if this is a public API change? Do we need a KIP for this?

**Comment by guozhangwang** [comment]
I do not think it is a public API: users do not need to change any code they have if they do not want to, as this change is for allowing them to get some "syntax sugar".

**Comment by joan38** [comment]
Ok I'm adding the Serialized, Consumed and Produced with implicit Serdes too

**Review by vvcephei — COMMENTED** [review]
Hey @joan38 , It looks good!

I think the sugar classes might benefit from scaladoc explaining what they are, what they do, and providing a few examples.

**Comment by vvcephei** [comment]
@guozhangwang, I agree that this is not changing or breaking any public APIs, but it seems like it is adding new public APIs. We can think of it as similar to adding new convenience methods to the existing builders. As such, it does seem like it needs a KIP.

**Comment by vvcephei** [comment]
@joan38 Also, you might already have this in the works, but I think we should keep up the trend of having test coverage on all these additions. 

We initially didn't have enough respect for how hard it is to just read a scala DSL and visually verify that it'll behave as expected. It might be over-learning, but I'd feel better if we could see the expected usage in some tests.

**Comment by joan38** [comment]
As you expected I wrote the tests 😄 and ported the documentation.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/package.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/package.scala]: You must have not compiled tests.
You can't import the type from java and the object from Scala if they have the same name but different package, unless you give the fully qualified name.
The type here does the import for you in the right package.

**Comment by joan38** [comment]
I forgot the `Joined` one

**Comment by joan38** [comment]
The KIP has been created:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-365%3A+Materialized%2C+Serialized%2C+Joined%2C+Consumed+and+Produced+with+implicit+Serde

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala
Comment (tedyu) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala]: What's the purpose for prefixing with kstream ?

If I remove the prefix, the code still compiles.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (tedyu) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: Please don't use individual's name.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (tedyu) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: It seems these two lines should be moved closer to line 31 where keySerde is documented

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Joined.scala
Comment (tedyu) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Joined.scala]: Drop author, please

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala]: Thanks, this was an IDE refator that introduced it.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: These are type param documentation. They are usually after the parameters.

Comment (tedyu) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: From KStream:
```
 * @tparam K Type of keys
 * @tparam V Type of values
 * @param inner The underlying Java abstraction for KStream
```

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: Why is that?
I asked that earlier and it seemed ok: https://github.com/apache/kafka/pull/5502#discussion_r210832920

Is there a special case for Alexis Seigneurin?:
https://github.com/apache/kafka/blob/trunk/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala#L3

Sorry I don't know much about those licensing.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Joined.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Joined.scala]: See reply above https://github.com/apache/kafka/pull/5551#discussion_r212813051

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: I guess that make sense to move them up since this is the order they are displayed in the signature. I will change that.

**Comment by joan38** [comment]
Retest this please

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: I'd recommend we adopt whatever are normal scaladoc conventions. I think I'm responsible for the KStream doc, but I wasn't aware of a convention.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: This change is missing from the KIP.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala:31]: thanks! I've been meaning to do this.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/JoinedTest.scala
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/JoinedTest.scala]: nit: you don't need the `()`, right?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java
Comment (joan38) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: Actually let's discuss about that.
Why `super` here?
I can add that in the KIP.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: I have wondered this many times myself.
This class was introduced in:
* https://github.com/apache/kafka/pull/3770
* https://cwiki.apache.org/confluence/display/KAFKA/KIP-182%3A+Reduce+Streams+DSL+overloads+and+allow+easier+use+of+custom+storage+engines

Interestingly, it's not a lower bound (it's exact) in the KIP.

I couldn't find any discussion of switching from exact to lower bound in the PR, the Jira, or the KIP.

There are many other method parameters that have lower-bound generics, so it was likely something that everyone had on their mind and didn't need to talk about. Or, it was just a carryover from the other methods that `Produced` replaced, and I just didn't dig deep enough.

I'm having a hard time imagining a situation in which the exact bound would be too tight. But of course my failure of imagination is not a proof.

Perhaps it's just that, which we do need to have an exact bound on the serdes, a lower bound is sufficient on the partitioner.

Maybe @guozhangwang @dguy @bbejeck or @mjsax can shed some light on it.

Your question is still valid, but just for the sake of argument, can you let us know why the bounded type is a problem for the sugared `Produced` ?

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: It's from KIP-100 https://cwiki.apache.org/confluence/display/KAFKA/KIP-100+-+Relax+Type+constraints+in+Kafka+Streams+API

Comment (joan38) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: @vvcephei if we keep the lower bound the compiler complains that Serdes need exact bounds for some reasons.

I think it makes sense since the type on `K` or `V` is supposed to be the same in the `Serde` and in the `StreamPartitioner`, under the `Produced` class. I just wonder why it works in Java.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: Thanks, @bbejeck !

@joan38 hmm. I'm not sure why it wouldn't work. All it's saying is that we want:
* a Serde for exactly K
* a Serde for exactly V
* a partitioner (function) that takes 1: anything that can be assigned to K and 2: anything that can be assigned to V

This seems like a perfectly reasonable set of constraints, and it accurately reflects our actual constraints:
* when we're serializing/deserializing, we need to have the exact right serde for the type in question, otherwise, we'll lose some information when we round-trip
* when we're picking a partition, we don't need exactly K and V, but we are processing a <K,V> document, so we need to be able to pass (k:K,v:V) arguments to the partition function. But if you want to write a more generic partition function that, say, picks partitions for any `Object` based on its `hashCode`, then this is fine by us.

I bet that there's some tricky type-bound scala syntax that will make it work. If you run out of ideas, I can try pulling the PR down and banging my head against it, too.

Comment (joan38) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: @vvcephei Let me try that again tonight. I keep you updated

**Review by bbejeck — COMMENTED** [review]
Thanks, @joan38 for the patch!  Overall LGTM but I have a meta-question regarding changes to `ProducedInternal` and `SerializedInternal` as they both now have full public access.  

Do we still need a `Produced` and `Serialized` and the internal versions of these two classes?  I believe we added the `XInternal` versions to restrict access to the internals of those classes.  I realize this question is out of scope for this PR, but I wanted to pose the question now as I'm thinking about it.

**Comment by tedyu** [comment]
If ProducedTest.scala is moved to the same package as ProducedInternal, then ProducedInternal can be package private instead of public, right ?

**Comment by tedyu** [comment]
Similar comment for SerializedTest.scala

File: streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java
Comment (joan38) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: Alright it just worked, so I think I was tired when I did it before

**Comment by joan38** [comment]
@tedyu Yes I'm not sure if we should move the tests in the `internal` package just for the sake of making it accessible?
@bbejeck To me it sounds like there should be no `XInternal` and just immutable ones but I'm a Scala guy talking here.
I don't know much how we can do to restrict this but still be able to use it freely in tests.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Produced.java]: :) an underappreciated component in programming

**Comment by vvcephei** [comment]
One problem with Java is that the access control mechanisms are very coarse. Let's look at ProducedInternal. The reason it exists is to expose the `keySerde()`, `valueSerde()`, and `streamPartitioner()` getters. The Produced class intentionally doesn't expose the getters.

Regardless of whether Produced is immutable, if we add these getters to `Produced`, they would become part of the public API (which needs a KIP, etc).

It doesn't really matter whether `ProducedInternal` has a public constructor or not, since it's still in the `internal` package, which documents that it's "not a public interface".

It's better, of course, if we can avoid making the constructor public, since then the compiler backs up our package-based "this is not public" statement.

If this were scala, we could add `private[streams]` modifiers to these methods and just put them in Produced, but there's no way to do this in Java... c'est la vie.

**Comment by joan38** [comment]
retest this please

**Comment by joan38** [comment]
Do you know why this is not triggering Jenkins jobs?

**Comment by joan38** [comment]
Retest this please

**Comment by joan38** [comment]
Test this please

**Comment by vvcephei** [comment]
I'm not sure...

Retest this, please.

**Comment by joan38** [comment]
It worked 👍

**Comment by vvcephei** [comment]
Yeah, Jenkins must have fallen asleep. (or maybe there were no available workers or something)

**Comment by vvcephei** [comment]
It looks like the bounds may be backwards:
```
00:04:36.049 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Produced.scala:60: type mismatch;
00:04:36.049  found   : org.apache.kafka.streams.processor.StreamPartitioner[_$1,_$2] where type _$2 <: V, type _$1 <: K
00:04:36.049  required: org.apache.kafka.streams.processor.StreamPartitioner[_ >: K, _ >: V]
00:04:36.049 Note: _$1 <: Any, but Java-defined trait StreamPartitioner is invariant in type K.
00:04:36.049 You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
00:04:36.049 Note: _$2 <: Any, but Java-defined trait StreamPartitioner is invariant in type V.
00:04:36.049 You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
00:04:36.049     ProducedJ.`with`(keySerde, valueSerde, partitioner)
```

**Comment by joan38** [comment]
Indeed, I will correct this.

**Comment by joan38** [comment]
@vvcephei I just corrected it and now see:
```
Error:(60, 22) type mismatch;
 found   : org.apache.kafka.common.serialization.Serde[K]
 required: org.apache.kafka.common.serialization.Serde[_$1] where type _$1 >: K
Note: K <: _$1, but Java-defined trait Serde is invariant in type T.
You may wish to investigate a wildcard type such as `_ <: _$1`. (SLS 3.2.10)
    ProducedJ.`with`(keySerde, valueSerde, partitioner)
```
Again, I don't understand how can we have `K` for part or the function and `_ >: K` for the other part of the function. To me it should be all on the same type.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala:42]: For my own education: why we only need `implicit` for the first parameter, instead of having it for both key and value serdes?

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
Comment (guozhangwang) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala]: Why need the `kstream.` prefix here?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala:42]: The implicit keyword applies to the whole set of parameters, not just to the first parameter.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
Comment (joan38) [reply streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala]: Wrong refactoring from my IDE. I will correct this.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (mjsax) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala]: The other code was release before it was merged into Apache Kafka. So the original copyright was with Alexis and Lightbend (ie, Lightbend copied Alexis code and later contributed it to AK).

For regular PRs, the contributor license agreement states, the ASF gets the copyright (from my understanding---also not a lawyer). Otherwise, it would also be a hassle to add copyright to any file that is changed... Also not a lawyer, but please read https://www.apache.org/licenses/icla.pdf (for committers it's required to file this -- for contributors consider the wiki information: https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes

```
When you contribute code, you affirm that the contribution is your original work and that you license the work to the project under the project's open source license. Whether or not you state this explicitly, by submitting any copyrighted material via pull request, email, or other means you agree to license the material under the project's open source license and warrant that you have the legal authority to do so.
```

**Comment by joan38** [comment]
retest this please

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Consumed.scala:42]: @guozhangwang you are right that the declaration syntax of implicit parameters in Scala is confusing.
See this proposal https://github.com/lampepfl/dotty/issues/1260 for Scala 3:
> The syntax (implicit x: T, y: U) is a bit strange in that implicit conceptually scopes over
x and y but looks like a modifier for just x.

That would make it clear that the `implicit` is applied to the whole group of curried parameters.

**Comment by joan38** [comment]
Are we good with this?

**Comment by vvcephei** [comment]
Hey @joan38 ,

You need to close the vote email thread and update the KIP status. I sent a message with the next steps to the vote thread for KIP-365.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ConsumedTest.scala
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ConsumedTest.scala:37]: Gah. I just now realized that's it's not ok to be caching these serdes. It's not part of your change, though. I'll send a separate PR to fix it.

Comment (vvcephei) [reply streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ConsumedTest.scala:37]: just FYI: https://issues.apache.org/jira/browse/KAFKA-7386 (I've assigned it to myself)

**Review by vvcephei — APPROVED** [review]
LGTM! Thanks!

We can merge as soon as the KIP is finalized. /cc @guozhangwang

**Comment by mjsax** [comment]
Btw: all KIPs should be backed by a Jira and the PR should have the ticket number in the title. Please update the PR and KIP accordingly after you created the Jira.

Also, this PR seems to need a rebase. Thx.

**Comment by joan38** [comment]
@mjsax @vvcephei All done.
Should we go ahead and merge?

**Pull request merged**
